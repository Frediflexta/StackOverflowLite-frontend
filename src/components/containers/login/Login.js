import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import { sendLoginRequest } from '../../../redux/actions/login/loginAction';
import Input from '../../presentation/Inputs/Inputs';
import Button from '../../presentation/button/Button';
import Notification from '../../presentation/notification/Notification';

class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      username: '',
      password: '',
      status: '',
      message: '',
      display: 'none',
    };
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  handleInputChange(event) {
    const { value } = event.target;
    this.setState({
      [event.target.name]: value
    });
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const {
      history, login
    } = this.props;

    try {
      await login({ ...this.state });
      this.setState({
        display: 'block',
        message: this.props.notification.message,
        status: this.props.notification.type,
      });
      if (this.props.notification.type === 'success') {
        return setTimeout(() => history.push('/', { prev: 'login' }), 500);
      }
    } catch (error) {
      /* do nothing */
    }
  }


  render() {
    const {
      username, password,
      status, message, display
    } = this.state;
    return (
      <section className="hero is-light is-fullheight">
        <div className="hero-body">
          <div className="container has-text-centered">
            <div className="column is-4 is-offset-4">
              <h3 className="title has-text-grey">Please login to proceed</h3>
              <div className="box">
                <form onSubmit={this.handleSubmit}>
                  <Notification
                    status={status}
                    message={message}
                    display={display}
                  />
                  <Input
                    fieldClassName="field"
                    controlClassName="control"
                    inputClassName="input is-large"
                    name="username"
                    type="username"
                    placeholder="Your Username"
                    value={username}
                    onChange={this.handleInputChange}
                    required
                  />

                  <Input
                    fieldClassName="field"
                    controlClassName="control"
                    inputClassName="input is-large"
                    name="password"
                    type="password"
                    placeholder="Your Password"
                    // Password must contain 6 or more characters
                    // a digit
                    // a lower-case letter
                    // an upper-case letter
                    pattern="(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,12}"
                    value={password}
                    onChange={this.handleInputChange}
                    required
                  />
                  <Button
                    className="button is-block is-info is-large is-fullwidth"
                    type="submit"
                    text="Login"
                  />
                </form>
              </div>
              <p className="has-text-grey">
                No account yet? <Link to="register">Sign Up</Link>
              </p>
            </div>
          </div>
        </div>
      </section>
    );
  }
}

Login.propTypes = {
  history: propTypes.object,
  login: propTypes.func,
  dispatch: propTypes.func,
  notification: propTypes.object,
};

const mapDispatchToProps = dispatch => ({
  login: user => dispatch(sendLoginRequest(user))
});

const mapStateToProps = state => ({
  notification: state.notificationReducer,
});

const LoginContainer = connect(mapStateToProps, mapDispatchToProps)(Login);
export default LoginContainer;
