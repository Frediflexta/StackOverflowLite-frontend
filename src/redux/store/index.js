import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/rootReducer';

// const middleware = process.env.NODE_ENV !== 'production'
//   ? [require('redux-immutable-state-invariant').default(), thunk]
//   : [thunk];

const middleware = [thunk];

if (process.env.NODE_ENV !== 'production') {
  const { createLogger } = require('redux-logger');
  const logger = createLogger({ collapsed: true });
  middleware.push(require('redux-immutable-state-invariant').default());
  middleware.push(logger);
}

/**
 * @param {*} initialState
 * @return {*} createStore
 */
function configureStore() {
  return createStore(
    rootReducer,
    applyMiddleware(...middleware),
  );
}

const store = configureStore();
window.store = store;

store.subscribe(() => localStorage.setItem('store', JSON.stringify(store.getState())));
export default store;
