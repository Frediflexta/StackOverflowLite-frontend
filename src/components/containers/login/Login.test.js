import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import Login from './Login';
import store from '../../../redux/store/index';

describe('<Login /> shallow rendering tests', () => {
  const props = {
    login: jest.fn(),
    history: { push: jest.fn() },
    notification: { type: 'success' },
  };
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <MemoryRouter>
        <Login {...props} store={store}/>
      </MemoryRouter>
    );
  });

  it('should render Login successfully', () => {
    expect(wrapper.find('form').exists()).toBe(true);
    // expect(wrapper.find('.login-intro').length).toBe(2);
    expect(wrapper.find('h3').text()).toBe('Please login to proceed');
  });

  it('matches the snapshot', () => {
    const tree = shallow(<Login {...props} store={store}/>);
    expect(tree).toMatchSnapshot();
  });

  it('should render all the children components successfully', () => {
    expect(wrapper.find('Notification').exists()).toBe(true);
    expect(wrapper.find('Input').exists()).toBe(true);
    expect(wrapper.find('Button').exists()).toBe(true);
  });

  it('should handle input changes successfully', () => {
    const instance = wrapper.find('Login').instance();
    const event = {
      target: {
        value: 'test',
        name: 'email'
      },
    };

    instance.handleInputChange(event);
    expect(instance.state.email).toEqual('test');
  });

  it('should handle submit form successfully', async () => {
    const instance = wrapper.find('Login').instance();
    const event = {
      preventDefault: jest.fn()
    };

    await instance.handleSubmit(event);
    // expect(instance.state.email).toEqual('test');
  });
});
