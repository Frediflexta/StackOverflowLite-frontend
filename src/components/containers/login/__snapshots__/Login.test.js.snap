// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> shallow rendering tests matches the snapshot 1`] = `
<Login
  history={
    Object {
      "push": [MockFunction],
    }
  }
  login={[Function]}
  notification={
    Object {
      "message": null,
      "type": null,
    }
  }
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
